### PR TITLE
[Misc] Add new Python backend for unit tests of funcs, coverage testing and debugging

### PR DESCRIPTION
**Pull request by hughperkins**
Issue: #

### Brief Summary

copilot:summary

### Walkthrough

copilot:walkthrough

### CODE REVIEW

File: python/quadrants/lang/ast/ast_transformers/call_transformer.py [hunk 1]
```diff
@@ -120,7 +120,7 @@ def _warn_if_is_external_func(ctx: ASTTransformerFuncContext, node):
     # origin input: 'qwerty {1} {} {1:.3f} {k:.4f} {k:}'.format(1.0, 2.0, k=k)
     # raw_string: 'qwerty {1} {} {1:.3f} {k:.4f} {k:}'
     # raw_args: [1.0, 2.0]
-    # raw_keywords: {'k': <ti.Expr>}
+    # raw_keywords: {'k': <qd.Expr>}
```
Comment (duburcqa) [inline python/quadrants/lang/ast/ast_transformers/call_transformer.py:123]: unrelated but ok

File: python/quadrants/lang/_ndarray.py [hunk 2]
```diff
@@ -280,9 +294,14 @@ class ScalarNdarray(Ndarray):
...
-            self.dtype, arr_shape, layout=Layout.NULL, zero_fill=True, dbg_info=_qd_core.DebugInfo(get_traceback())
-        )
+        if impl.is_python_backend():
+            assert torch_is_available
+            torch_dtype = dtype_to_torch_dtype(dtype)
+            self.arr = torch.zeros(shape=arr_shape, dtype=torch_dtype)
+        else:
+            self.arr = impl.get_runtime().prog.create_ndarray(
+                self.dtype, arr_shape, layout=Layout.NULL, zero_fill=True, dbg_info=_qd_core.DebugInfo(get_traceback())
+            )
```
Comment (duburcqa) [inline python/quadrants/lang/_ndarray.py]: I would recommend to remove guarded torch import, and move it locally inside branch `if impl.is_python_backend():` directly.

File: python/quadrants/lang/_ndrange.py [hunk 3]
```diff
@@ -5,24 +5,37 @@
...
 from quadrants.lang.expr import Expr
 from quadrants.lang.matrix import Matrix
 from quadrants.types.utils import is_integral
 
 
+def _coerce_to_int(v):
+    """Convert 0-d tensors to int for ndrange bounds in python backend."""
+    if torch is not None and isinstance(v, torch.Tensor) and v.ndim == 0:
+        return int(v.item())
+    return v
```
Comment (duburcqa) [inline python/quadrants/lang/_ndrange.py]: Same. I think it makes more sense to move torch import inside the functions where it is used.

File: python/quadrants/lang/impl.py [hunk 4]
```diff
@@ -71,6 +74,14 @@
...
 )
 
+from . import _py_tensor as py_tensor
+
+torch = None
+try:
+    import torch
+except Exception:
+    pass
+
```
Comment (duburcqa) [inline python/quadrants/lang/impl.py]: This block is useless here.

File: python/quadrants/lang/impl.py [hunk 5]
```diff
@@ -1223,6 +1267,10 @@ def current_cfg():
     return get_runtime().prog.config()
 
 
+def is_python_backend() -> bool:
+    return get_runtime()._arch == _qd_core.Arch.python
```
Comment (duburcqa) [inline python/quadrants/lang/impl.py]: `_qd_core.Arch.python` should be replaced by a proxy, unless efficiency does not matter.

File: python/quadrants/lang/kernel.py [hunk 6]
```diff
@@ -302,6 +304,7 @@ def __init__(self, _func: Callable, autodiff_mode: AutodiffMode, _is_classkernel
         self.launch_observations = LaunchObservations()
 
         self.launch_context_buffer_cache = LaunchContextBufferCache()
+        self.arch_is_python: bool | None = None
```
Comment (duburcqa) [inline python/quadrants/lang/kernel.py]: This flag is unused.

File: python/quadrants/lang/matrix.py [hunk 7]
```diff
@@ -249,6 +267,13 @@ class Matrix(QuadrantsOperations):
     _is_matrix_class = True
     __array_priority__ = 1000
 
+    def __new__(cls, arr, dt=None):
+        if impl.is_python_backend():
+            assert torch is not None
```
Comment (duburcqa) [inline python/quadrants/lang/matrix.py]: This check is useless.

File: python/quadrants/lang/_py_tensor.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    except Exception:
+        pass
+
+    if torch is not None:
+        _TensorBase = torch.Tensor
+    else:
+        _TensorBase = object
+
+
+class MyTorchTensor(_TensorBase):
```
Comment (duburcqa) [inline python/quadrants/lang/_py_tensor.py]: This is not a correct name.

File: python/quadrants/lang/_py_tensor.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    torch = None
+    try:
+        import torch
+    except Exception:
+        pass
+
+    if torch is not None:
+        _TensorBase = torch.Tensor
+    else:
+        _TensorBase = object
```
Comment (duburcqa) [inline python/quadrants/lang/_py_tensor.py]: I don't think it should be possible to import this file without torch installed. It suggest to delegate the responsibility to make sure torch is available to the other modules importing this one.

File: python/quadrants/lang/_py_tensor.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    def x(self):
+        return super().__getitem__(0)
+
+    @property
+    def y(self):
+        return super().__getitem__(1)
+
+    @property
+    def z(self):
+        return super().__getitem__(2)
```
Comment (duburcqa) [inline python/quadrants/lang/_py_tensor.py]: just do `return super()[0]`

File: python/quadrants/lang/_py_tensor.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+                if isinstance(item, float) and item.is_integer():
+                    item = int(item)
+                yield item
+            else:
+                yield val
+
+    def __getitem__(self, key):  # type: ignore[override]
+        key = MyTorchTensor._unpack_key(key)
+        if not isinstance(key, tuple) and key == 0 and self.size() == ():
+            return self.item()
```
Comment (duburcqa) [inline python/quadrants/lang/_py_tensor.py:94]: I don't understand why this is needed.

File: python/quadrants/lang/matrix.py [hunk 12]
```diff
@@ -36,9 +37,26 @@
 from quadrants.types.enums import Layout
 from quadrants.types.utils import is_signed
 
+from . import _py_tensor as py_tensor
+
+torch = None
+try:
+    import torch
+except Exception:
+    pass
```
Comment (duburcqa) [inline python/quadrants/lang/matrix.py]: I would rather switch to local import

File: python/quadrants/lang/misc.py [hunk 13]
```diff
@@ -716,6 +742,7 @@ def is_arch_supported(arch):
         metal: _qd_core.with_metal,
         vulkan: _qd_core.with_vulkan,
         cpu: lambda: True,
+        python: lambda: True,
```
Comment (duburcqa) [inline python/quadrants/lang/misc.py]: This backend should not be available if torch is not installed.

File: python/quadrants/lang/struct.py [hunk 14]
```diff
@@ -501,7 +511,8 @@ def fill(self, val):
 
     def _initialize_host_accessors(self):
         for v in self._members:
-            v._initialize_host_accessors()
```
Comment (duburcqa) [inline python/quadrants/lang/struct.py:504]: I would rather have `_initialize_host_accessors` a no-op for python backend

File: python/quadrants/lang/util.py [hunk 15]
```diff
@@ -41,9 +47,7 @@ def has_pytorch():
     _env_torch = os.environ.get("QD_ENABLE_TORCH", "1")
     if not _env_torch or int(_env_torch):
         try:
-            import torch  # noqa: F401 pylint: disable=C0415
-
-            _has_pytorch = True
+            _has_pytorch = torch is not None
         except:
             pass
     return _has_pytorch
```
Comment (duburcqa) [inline python/quadrants/lang/util.py]: I'm not convinced this new pattern is better.

File: python/quadrants/lang/_py_tensor.py (hunk 8 above)
Comment (hughperkins) [reply python/quadrants/lang/_py_tensor.py]: thoughts on a better name?

File: python/quadrants/lang/_py_tensor.py (hunk 11 above)
Comment (hughperkins) [reply python/quadrants/lang/_py_tensor.py:94]: added comment to th ecode.

**Comment by hughperkins** [comment]
nwe review from opus-4.6:

Overview

  This branch adds a pure-Python backend to Quadrants. When you do qd.init(qd.python), kernels and functions execute as plain Python — no C++ compilation, no GPU needed.
  Fields and ndarrays are backed by PyTorch tensors instead of the C++ runtime.
  22 files changed, +1127 / -30 lines across C++, Python, and tests.

  ────────────────────────────────────────



  Architecture

  The design is clean — almost all changes are guarded by is_python_backend() checks that short-circuit into Python-native paths:
  • `Kernel.__call__` / `Func.__call__`: When on the python backend, just call self.func(*py_args, **kwargs) directly — skip compilation, AST transforms, all of it.
  • `qd.field()` / `qd.ndarray()`: Return PyTensor instances (torch tensors) instead of C++ data structures.
  • Atomics (atomic_add, atomic_sub, etc.): Become simple read-modify-write, which is correct since the backend is single-threaded.
  • `sync()`, `loop_config()`: Become no-ops.
  • C++ side: Minimal — just adds python to the arch enum, sets program_impl_ = nullptr in the Program constructor, and skips finalization. Very well contained.


  ────────────────────────────────────────



  PyTensor

  The central new abstraction is _py_tensor.PyTensor, a torch.Tensor subclass. Key design choices:
  • `_batch_shape`: Separates batch dims from element dims. A vec3 field of shape (N,) has torch size (N, 3) but .shape returns (N,). This lets kernel code like a.shape[0]
    work transparently.
  • `__torch_function__`: Unwraps PyTensor when mixed with foreign tensor subclasses to avoid dispatch issues.
  • `__getattr__`: Delegates unknown attributes (.norm(), .dot(), etc.) to matrix_ops, so Quadrants math APIs work on these tensors.
  • `_tc` / `_T_tc` / `_np` / `_T_np`: Mimics the Field/Ndarray view convention so downstream code that accesses these attributes works unchanged.


  ────────────────────────────────────────



  Potential Concerns

  1. Monkey-patching `DataTypeCxx.__call__`: _install_python_backend_dtype_call() patches the C++ DataType class globally so qd.f32(0.0) returns a Python float. There's a
     guard against double-install, which is good. But it's permanent global mutation — if you later qd.init(qd.cpu), the patched method checks is_python_backend() at call
     time, so it should still work, but it's fragile. If the C++ side ever changes how __call__ works, this could silently mask it.
  2. Overriding `.shape` on a torch.Tensor subclass: This is playing with fire. A lot of PyTorch internals and third-party code assumes .shape equals the actual tensor
     dimensions. The __torch_function__ unwrapping helps, but any code path that bypasses __torch_function__ (C++ torch internals, some serialization paths) could see
     mismatched shape/size.
  3. `cast()` uses string matching: The python backend path in cast() does "float" in str(dtype) or "f16" in str(dtype) — this is fragile. If dtype string representations
     change or new types are added, it could misclassify.
  4. `_coerce_to_int` in ndrange: Needed because some code paths produce 0-d tensors instead of plain ints for loop bounds. This is a symptom of the .shape override
     leaking — it works, but indicates the abstraction isn't fully sealed.
  5. `StructField._register_fields()` guard: Now conditioned on all(isinstance(v, Field) ...), silently skipping registration for PyTensors. Works correctly, but is an
     implicit contract.
  6. `grouped(x)` returns [[i] for i in range(x.shape[0])] — I'd want to verify this matches the semantics of the grouped iterator on other backends. It looks like it
     converts to a per-element iteration rather than actual grouping.

File: python/quadrants/lang/_py_tensor.py (hunk 11 above)
Comment (duburcqa) [reply python/quadrants/lang/_py_tensor.py:94]: Why not just return self?

File: python/quadrants/lang/_py_tensor.py [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+
+def _setup_views(tensor, batch_ndim):
+    """Set _tc/_T_tc/_np/_T_np instance attributes matching the qd.Field/Ndarray convention."""
+    tensor._tc = tensor
+    tensor._T_tc = tensor.movedim(batch_ndim - 1, 0) if batch_ndim > 1 else tensor
+    tensor._np = tensor.numpy()
+    tensor._T_np = tensor._T_tc.numpy()
+    tensor._batch_shape = tensor.size()[:batch_ndim]
+    tensor.grad = None
+    tensor.dual = None
```
Comment (duburcqa) [inline python/quadrants/lang/_py_tensor.py]: Why not defining this attributes at init? Instead of hacking your own class?

Comment (hughperkins) [reply python/quadrants/lang/_py_tensor.py]: here is what cursor said:

Screenshot 2026-03-08 at 16 12 33
Screenshot 2026-03-08 at 16 12 38
Screenshot 2026-03-08 at 16 13 38

File: python/quadrants/lang/_py_tensor.py (hunk 11 above)
Comment (hughperkins) [reply python/quadrants/lang/_py_tensor.py:94]: cursor said:

Screenshot 2026-03-08 at 16 17 37
